Tolerate modules in same directory

When a directory has multiple project model files, e.g.

project/pom.xml
project/parent.xml

and a change is made in the directory, instead of throwing 

`Exception during gitflow-incremental-builder execution occurred. Duplicate key MavenProject`

assume the change applies to both

Yeah its because one of the projects is a Maven plugin required by the others and I just don't want to have that extra directory so high up in the tree. Perfectly valid though! :)

```
[ERROR] Exception during gitflow-incremental-builder execution occurred. Duplicate key MavenProject: com.toreador:toreador-maven-parent:2.2.0 @ /git/project/parent.xml -> [Help 1]
org.apache.maven.MavenExecutionException: Exception during gitflow-incremental-builder execution occurred.
    at com.vackosar.gitflowincrementalbuild.boundary.MavenLifecycleParticipant.perform (MavenLifecycleParticipant.java:86)
    at com.vackosar.gitflowincrementalbuild.boundary.MavenLifecycleParticipant.afterProjectsRead (MavenLifecycleParticipant.java:64)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:264)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.IllegalStateException: Duplicate key MavenProject: com.toreador:toreador-maven-parent:2.2.0 @ /git/project/parent.xml
    at java.util.stream.Collectors.lambda$throwingMerger$0 (Collectors.java:133)
    at java.util.HashMap.merge (HashMap.java:1254)
    at java.util.stream.Collectors.lambda$toMap$58 (Collectors.java:1320)
    at java.util.stream.ReduceOps$3ReducingSink.accept (ReduceOps.java:169)
    at java.util.ArrayList$ArrayListSpliterator.forEachRemaining (ArrayList.java:1380)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:481)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:471)
    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential (ReduceOps.java:708)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.collect (ReferencePipeline.java:499)
    at com.vackosar.gitflowincrementalbuild.control.Modules.createPathMap (Modules.java:19)
    at com.vackosar.gitflowincrementalbuild.control.ChangedProjects.get (ChangedProjects.java:32)
    at com.vackosar.gitflowincrementalbuild.boundary.UnchangedProjectsRemover.doAct (UnchangedProjectsRemover.java:103)
    at com.vackosar.gitflowincrementalbuild.boundary.UnchangedProjectsRemover.act (UnchangedProjectsRemover.java:66)
    at com.vackosar.gitflowincrementalbuild.boundary.MavenLifecycleParticipant.perform (MavenLifecycleParticipant.java:79)
    at com.vackosar.gitflowincrementalbuild.boundary.MavenLifecycleParticipant.afterProjectsRead (MavenLifecycleParticipant.java:64)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:264)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)     
```

--- a/pom.xml
+++ b/pom.xml
@@ -639,7 +639,7 @@
                                         <limit>
                                             <counter>BRANCH</counter>
                                             <value>COVEREDRATIO</value>
-                                            <minimum>0.84</minimum>
+                                            <minimum>0.85</minimum>
                                         </limit>
                                         <limit>
                                             <counter>CLASS</counter>

--- a/src/main/java/com/vackosar/gitflowincrementalbuild/control/ChangedProjects.java
+++ b/src/main/java/com/vackosar/gitflowincrementalbuild/control/ChangedProjects.java
@@ -3,8 +3,9 @@ package com.vackosar.gitflowincrementalbuild.control;
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.Collections;
+import java.util.List;
 import java.util.Map;
-import java.util.Objects;
 import java.util.Set;
 import java.util.stream.Collectors;
 
@@ -37,14 +38,13 @@ public class ChangedProjects {
     @Inject private Modules modules;
 
     public Set<MavenProject> get(Configuration config) throws GitAPIException, IOException {
-        Map<Path, MavenProject> modulesPathMap = modules.createPathMap(config.mavenSession);
+        Map<Path, List<MavenProject>> modulesPathMap = modules.createPathMap(config.mavenSession);
         return differentFiles.get(config).stream()
-                .map(path -> findProject(path, modulesPathMap))
-                .filter(Objects::nonNull)
+                .flatMap(path -> findProject(path, modulesPathMap).stream())
                 .collect(Collectors.toSet());
     }
 
-    private MavenProject findProject(Path diffPath, Map<Path, MavenProject> modulesPathMap) {
+    private List<MavenProject> findProject(Path diffPath, Map<Path, List<MavenProject>> modulesPathMap) {
         // Strip src/* subpath (if present) to make sure that embedded (test) projects contribute
         // to the "change state" of containing reactor module instead of considering them as separate (non-reactor) modules.
         Path path = stripSrcSubpath(diffPath);
@@ -55,19 +55,21 @@ public class ChangedProjects {
         }
         if (path == null) {
             logger.warn("Ignoring changed file outside build project: {}", diffPath);
-            return null;
+            return Collections.emptyList();
         }
-        MavenProject changedReactorProject = modulesPathMap.get(path);
-        if (changedReactorProject == null) {
+        List<MavenProject> changedReactorProjects = modulesPathMap.get(path);
+        if (changedReactorProjects == null) {
             logger.warn("Ignoring changed file in non-reactor module: {}", diffPath);
-            return null;
+            return Collections.emptyList();
         }
         logger.debug("Changed file: {}", diffPath);
-        Boolean testOnlyFlag = (Boolean) changedReactorProject.getContextValue(CTX_TEST_ONLY);
-        if (!Boolean.FALSE.equals(testOnlyFlag)) {
-            changedReactorProject.setContextValue(CTX_TEST_ONLY, diffPath.startsWith(path.resolve("src").resolve("test")));
+        for (MavenProject changedReactorProject : changedReactorProjects) {
+            Boolean testOnlyFlag = (Boolean) changedReactorProject.getContextValue(CTX_TEST_ONLY);
+            if (!Boolean.FALSE.equals(testOnlyFlag)) {
+                changedReactorProject.setContextValue(CTX_TEST_ONLY, diffPath.startsWith(path.resolve("src").resolve("test")));
+            }
         }
-        return changedReactorProject;
+        return changedReactorProjects;
     }
 
     @SuppressFBWarnings(value = "NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE",

--- a/src/main/java/com/vackosar/gitflowincrementalbuild/control/Modules.java
+++ b/src/main/java/com/vackosar/gitflowincrementalbuild/control/Modules.java
@@ -1,8 +1,11 @@
 package com.vackosar.gitflowincrementalbuild.control;
 
 import java.nio.file.Path;
+import java.util.Collections;
+import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import javax.inject.Named;
 import javax.inject.Singleton;
@@ -14,9 +17,16 @@ import org.apache.maven.project.MavenProject;
 @Named
 public class Modules {
 
-    public Map<Path, MavenProject> createPathMap(MavenSession session) {
-        return session.getAllProjects().stream()
-                .collect(Collectors.toMap(Modules::getPath, project -> project));
+    /**
+     * Returns all projects of the given session, mapped by their project base directory.
+     * In very special setups, multiple modules can exists in the same directory and so the mapping value is a list of projects.
+     */
+    public Map<Path, List<MavenProject>> createPathMap(MavenSession session) {
+        return session.getAllProjects().stream().collect(
+                Collectors.toMap(
+                        Modules::getPath,
+                        Collections::singletonList,
+                        (l1, l2) -> Stream.concat(l1.stream(), l2.stream()).collect(Collectors.toList())));
     }
 
     private static Path getPath(MavenProject project) {

--- a/src/test/java/com/vackosar/gitflowincrementalbuild/control/ModulesTest.java
+++ b/src/test/java/com/vackosar/gitflowincrementalbuild/control/ModulesTest.java
@@ -1,0 +1,69 @@
+package com.vackosar.gitflowincrementalbuild.control;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.data.MapEntry.entry;
+import static org.mockito.Mockito.when;
+
+import java.io.File;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.project.MavenProject;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class ModulesTest {
+
+    @Mock
+    private MavenSession sessionMock;
+
+    private final Modules underTest = new Modules();
+
+    @Test
+    void none() {
+        assertThat(underTest.createPathMap(sessionMock)).isEmpty();
+    }
+
+    @Test
+    void single() {
+        MavenProject projMock = mockProject("proj");
+        when(sessionMock.getAllProjects()).thenReturn(Arrays.asList(projMock));
+
+        assertThat(underTest.createPathMap(sessionMock)).containsOnly(
+                entry(Paths.get("proj").toAbsolutePath(), Arrays.asList(projMock)));
+    }
+
+    @Test
+    void multiple() {
+        MavenProject projMock1 = mockProject("proj1");
+        MavenProject projMock2 = mockProject("proj2");
+        when(sessionMock.getAllProjects()).thenReturn(Arrays.asList(projMock1, projMock2));
+
+        assertThat(underTest.createPathMap(sessionMock)).containsOnly(
+                entry(Paths.get("proj1").toAbsolutePath(), Arrays.asList(projMock1)),
+                entry(Paths.get("proj2").toAbsolutePath(), Arrays.asList(projMock2)));
+    }
+
+    @Test
+    void multiple_twoWithSameDir() {
+        MavenProject projMock1 = mockProject("proj1");
+        MavenProject projMock2 = mockProject("proj2+3");
+        MavenProject projMock3 = mockProject("proj2+3");
+        when(sessionMock.getAllProjects()).thenReturn(Arrays.asList(projMock1, projMock2, projMock3));
+
+        assertThat(underTest.createPathMap(sessionMock)).containsOnly(
+                entry(Paths.get("proj1").toAbsolutePath(), Arrays.asList(projMock1)),
+                entry(Paths.get("proj2+3").toAbsolutePath(), Arrays.asList(projMock2, projMock3)));
+    }
+
+
+    public MavenProject mockProject(String dirName) {
+        MavenProject projMock = Mockito.mock(MavenProject.class);
+        when(projMock.getBasedir()).thenReturn(new File(dirName));
+        return projMock;
+    }
+}

